runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -9472,29 +9478,48 @@ static TR::Register* inlineStringHashCode(TR::Node* node, bool isCompressed, TR:
    auto index = cg->allocateRegister();
    auto hash = cg->allocateRegister();
    auto tmp = cg->allocateRegister();
+   auto loopLimit = cg->allocateRegister();
    auto hashXMM = cg->allocateRegister(TR_VRF);
    auto tmpXMM = cg->allocateRegister(TR_VRF);
    auto multiplierXMM = cg->allocateRegister(TR_VRF);
 
    auto begLabel = generateLabelSymbol(cg);
    auto endLabel = generateLabelSymbol(cg);
    auto loopLabel = generateLabelSymbol(cg);
+   auto bigLoopLabel = generateLabelSymbol(cg);
+   auto doneLabel = generateLabelSymbol(cg);
    begLabel->setStartInternalControlFlow();
    endLabel->setEndInternalControlFlow();
    auto deps = generateRegisterDependencyConditions((uint8_t)6, (uint8_t)6, cg);
    deps->addPreCondition(address, TR::RealRegister::NoReg, cg);
    deps->addPreCondition(index, TR::RealRegister::NoReg, cg);
-   deps->addPreCondition(length, TR::RealRegister::NoReg, cg);
+   deps->addPreCondition(loopLimit, TR::RealRegister::NoReg, cg);
    deps->addPreCondition(multiplierXMM, TR::RealRegister::NoReg, cg);
    deps->addPreCondition(tmpXMM, TR::RealRegister::NoReg, cg);
    deps->addPreCondition(hashXMM, TR::RealRegister::NoReg, cg);
    deps->addPostCondition(address, TR::RealRegister::NoReg, cg);
    deps->addPostCondition(index, TR::RealRegister::NoReg, cg);
-   deps->addPostCondition(length, TR::RealRegister::NoReg, cg);
+   deps->addPostCondition(loopLimit, TR::RealRegister::NoReg, cg);
    deps->addPostCondition(multiplierXMM, TR::RealRegister::NoReg, cg);
    deps->addPostCondition(tmpXMM, TR::RealRegister::NoReg, cg);
    deps->addPostCondition(hashXMM, TR::RealRegister::NoReg, cg);
 
+   // Generate Main Loop; 4x Unrolled seems to yield the best performance for large arrays
+#ifdef TR_TARGET_64BIT
+   static char *unrollVar = feGetEnv("TR_setInlineStringHashCodeUnrollCount");
+   int32_t unrollCount = unrollVar ? atoi(unrollVar) : 4;
+#else
+   int32_t unrollCount = 1;
+#endif
+
+   int32_t charsPerMainLoopIteration = unrollCount * vectorSizes[vl - TR::VectorLength128];
+
+   generateRegRegInstruction(TR::InstOpCode::XORRegReg(), node, index, index, cg);
+   generateRegRegInstruction(TR::InstOpCode::XORRegReg(), node, hash, hash, cg);
+   generateRegRegInstruction(TR::InstOpCode::MOV4RegReg, node, loopLimit, length, cg);
+   generateRegImmInstruction(TR::InstOpCode::AND4RegImm4, node, loopLimit, charsPerMainLoopIteration - 1, cg);
```

Comment:
This only works accurately if `charsPerMainLoopIteration` is a power of 2 (I think).  Maybe add an assert in the code here to check for that.  Practically speaking, I think this can only happen if someone specifies an `unrollCount` that is not a power of 2.